### ISSUE TITLE
Roadmap

### ISSUE THREAD

**Issue by jondot** [jondot/hygen#31]
Hello everyone,

I've decided to aggregate popular feature requests and track/indicate their status. Please feel free to use this thread to discuss these and most importantly - generate great ideas.

- [x] **Load global (custom) helpers [issue](https://github.com/jondot/hygen/issues/21)**

Done! see more here http://www.hygen.io/extensibility

- [x] **Recursively list and execute generators [issue](https://github.com/jondot/hygen/issues/24)**

Lacking a performant solution, we're opting to a best-practice [contributed](https://github.com/jondot/hygen/issues/24) by @jfols. See official documentation [here](http://www.hygen.io/faq/#i-want-to-use-generators-from-a-single-place)

- [x] **dynamic prompt generation triggered by prompt input [issue](https://github.com/jondot/hygen/issues/17)** thanks: @jaykoontz

**Comment by jaykoontz** [comment]
Dynamic prompt generation is possible with inquirer. If you look at this example: https://github.com/SBoudrias/Inquirer.js/blob/master/examples/recursive.js, given access to inquirer.prompt, we can call inquirer.prompt again.

In hygen, though, the call to prompt is hardcoded:
```
const prompt = (actionfolder: string) => {
  const promptfile = path.join(actionfolder, 'prompt.js')
  if (!fs.existsSync(promptfile)) {
    return Promise.resolve({})
  }
  // lazy loads inquirer (80ms load time)
  // $FlowFixMe
  return require('inquirer').prompt(require(promptfile))
}
```

You could have this use a callback that's given a reference to prompt, but coded by the user instead.

**Comment by jaykoontz** [comment]
And I don't know if this relates to adding custom helpers, but one thing I've had to rely (but is a massive code smell) is using global vars to help me with file path names in the frontmatter.

To reduce my prompt questions, I took one of the inputs (ex: GET_NOTE_AUTHORS) and transformed + broke it out into two new variables: verb (get), and fileName (getNoteAuthors) so I could place it in the frontmatter like so:
```
---
  to: api/<%= verb %>/<%= fileName %>API.js
---
```

It would've been nice to have one file I could count on to transform all my global vars, because it's hard to transform them through inquirer in all the ways that I want.

**Comment by jondot** [comment]
@jaykoontz thanks for these awesome ideas! 🥇  

You've just demolished a painful roadblock regarding inquirer, I can't wait to implement this! :)

You second idea is super useful, right now I'm thinking about a special 'params.js' file, where you can do all of your preprocessing tasks. I'll have to spike that and see how it feels, versus global custom helpers, or maybe a combination of the two. Super interesting!

**Comment by jaykoontz** [comment]
Great! And that'd be wonderful. I'll keep an eye out for it, I think global custom helpers would be great since sometimes I want more control in other areas as well (it can be hard to inject code after the final import in a js file, for instance), and global custom helpers could be a good way around things.

**Comment by jondot** [comment]
@jaykoontz done! check out  http://www.hygen.io/generators#advanced-prompts

**Comment by jaykoontz** [comment]
@jondot, that's great, I'll give it a run!

**Comment by jondot** [comment]
@jaykoontz so, advanced prompts became... "advanced params" http://www.hygen.io/generators#advanced-params

The "upgrade" in the way prompting are made, and a previous PR that makes sure arguments are handed over to prompting, opened the door to build params functions:

```
//prompt.js
module.exports = {
   params: ({inquirer, args})=>{
      return {newParam: args.oldParam}
  }
}
```
Note the new function is called `params` but the file itself is _still_ called `prompt.js`, since I want changes to be backwards compatible (as much as possible). 

I think in the near future it will just be `index.js` in each folder, and this index file will export many functions, each dealing with its own concern. Either way I hope to support both _prompt.js_ file for the simplest use case of a 'dumb' prompt, and `index.js` file for the advanced use cases of custom logic and prompting.

**Comment by jaykoontz** [comment]
That's perfect, prompt.js is much better for importing utility functions than the ejs.t files. 

I'll refactor with that change and see how it plays out, looks promising

**Comment by jaykoontz** [comment]
Yep, everything works as expected now. Opening up inquirer allows me to insert my own custom global variables based off previous answers. Great stuff, thanks @jondot:
```
 return inquirer.prompt(questions).then(modifier.modifyPromptAnswers)
```

**Comment by jondot** [comment]
Update: we're now supporting global helpers and extensibility via a `.hygen.js` file: see here http://www.hygen.io/extensibility

**Comment by Lughino** [comment]
Hello @jondot, awesome job with this tool!
I'm starting to use it in my project and I found a couple of suggestions in order to have more flexibility.
I have different project in different teams, we use the same structure cross teams and having the same template in every project it's a bit annoying.
I already use a separate package adding it with `hygen-add`, but the problem still exists.. `hygen-add` copy the templates into the project and when we change the structure in the separate template we need to update the `_templates` folder in each project, or relaunch `hygen-add` in each project.
Also we have the problem to launch the generator only in the root of the project.
So, in order to keep the same template should be great if the `hygen-add` add the template to the global.
This solve different problems:
- we are not limited anymore to launch the generator in the root of the project, we can generate new files in every folder!
- the plugin can be updated with a command like `hygen plugin update` and scheduling every x days when we launch any `hygen` command, in this way we update the template in only one single place. Also the templates are already versioned in another repository.

**Comment by jondot** [comment]
Hi Luca, welcome!

I love your feedback, keep it coming!

I’m brainstorming on letting hygen discover the closest template directory from anywhere so your feedback comes at a fantastic timing. Consider it solved in a couple days.

Regarding hygen-add and template copying, the original thoughts was copying is better than trying to figure out a “smart” sharing and versioning strategy. I took the same approach the Go team took - do the simplest thing possible until there is enough requests for a different thing.

Actually with the new .hygen.js customization you can point to any templates folder you like. This actually solves point (1) and (2) out of the box.

I’ll drop a concrete example here in a couple hours (on mobile currently)

**Comment by jondot** [comment]
Luca,

Here's how you can solve it today, with the current Hygen version:

1. In the root of your project, create a `.hygen.js` file
2. In the `.hygen.js` file, place this content:

```javascript
module.exports = {
  templates: `${__dirname}/<path to your templates directory>/_templates`
}
```

Now you can run Hygen from _any_ subfolder in your project, and it will pick the same templates directory.

Does that work for you?

**Comment by Lughino** [comment]
Hi Dotan,
Thank you so much for the quick answer!
Yes this as temporary solution fit perfectly with the needs, reducing at
minimum the duplication of templates. And the update can be executed
manually in the templates folder.
This unblock totally the initial use of the generator cross teams.

**Comment by jondot** [comment]
Made my day! So happy to hear!
Of course if I end up with a better experience I’ll update as soon as possible (Ideally I’d like to make this experience automatic without a js file)

**Comment by jondot** [comment]
Closing up this ad-hoc roadmap, thanks all for motivating me to do all these great additions!
❤️
